### PR TITLE
Enable user configurability of jet reconstruction

### PR DESCRIPTION
**Pull request by ruse-traveler**
### Briefly, what does this PR introduce?

This PR updates the interfaces to jet reconstruction algorithm so as to allow users to change parameters of the algorithm from the command line, via a configuration file, or through initializer lists. For example:

`eicrecon -Preco:ReconstructedJets:minCstPt=2`

will raise the minimum pT of objects fed to the jet algorithm to 2 GeV/c for the reconstructed jet collection.

The utilized jet algorithm, recombination scheme, and area type can also be set in the same manner. If any of these are set to an unknown value (e.g. `-Preco:ReconstructedJets:jetAlgo=mystery_algo`), the EICrecon algorithm will issue a warning and run with a default value instead.

### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [X ] New feature (issue #901 )
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [X ] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users need to make to their code?

No.

### Does this PR change default behavior?

No.

### CODE REVIEW

File: src/algorithms/reco/JetReconstruction.cc [hunk 1]
```diff
@@ -4,26 +4,44 @@
...
+      m_cfg.jetAlgo = m_defaultFastjetOpts.jetAlgo;
+    }
+    if (m_mapRecombScheme.find(m_cfg.recombScheme) == m_mapRecombScheme.end()) {
+      m_log->warn(" Unknown recombination scheme \"{}\" specified! Setting scheme to default and proceeding.", m_cfg.recombScheme);
+      m_cfg.recombScheme = m_defaultFastjetOpts.recombScheme;
+    }
+    if (m_mapAreaType.find(m_cfg.areaType) == m_mapAreaType.end()) {
+      m_log->warn(" Unknown area type \"{}\" specified! Setting type to default and proceeding.", m_cfg.areaType);
+      m_cfg.areaType = m_defaultFastjetOpts.areaType;
+    }
```
Comment (veprbl) [inline src/algorithms/reco/JetReconstruction.cc:39]: If user's choice of option is not found, don't replace it with something else. Throw an exception (we commonly use JException), let it complain loudly.

**Comment by bspage912** [comment]
I will try to look at this tomorrow, but I have a question on the expected behavior. If I execute the command you list above "eicrecon -Preco:ReconstructedJets:minCstPt=2", does this modify the existing branches created using  ReconstructedJets_factory - currently ReconstructedJets and ReconstructedChargedJets, or does it create a new branch(es) with this specific change? One could for example think of wanting to generate a separate branches, each with a different R, for comparisons. Does this PR allow that functionality?

**Comment by veprbl** [comment]
@bspage912 The :ReconstructedJets: part determines which "factory tag" (collection name) you are setting options for. Different radii can be setup in the arguments to the factory generator in reco.cc.

File: src/algorithms/reco/JetReconstruction.cc (hunk 1 above)
Comment (ruse-traveler) [reply src/algorithms/reco/JetReconstruction.cc:39]: On 2nd thought: I like that approach. My 1st thought was to keep it "not so intrusive", but if a user is actively specifying something other than the default configuration, then they probably want loud complaints. I'll swap out the warns for errors and have it throw an exception!

**Comment by ruse-traveler** [comment]
> I will try to look at this tomorrow, but I have a question on the expected behavior. If I execute the command you list above "eicrecon -Preco:ReconstructedJets:minCstPt=2", does this modify the existing branches created using ReconstructedJets_factory - currently ReconstructedJets and ReconstructedChargedJets, or does it create a new branch(es) with this specific change? One could for example think of wanting to generate a separate branches, each with a different R, for comparisons. Does this PR allow that functionality?

Exactly as @veprbl said! That command will change the min cst. pt for the existing reconstructed jets collection. And while I don't think we have any tools in place yet that would dynamically generate a new branch in events based on commands like that one, this PR does make it easier for a user to add their own factory!

Previously, since the parameters were hard coded into the algorithm, all factories using the `JetReconstruction` algorithm had to use those. With this PR, we can set the parameters (preferably via initializer lists in `reco.cc`, but also possibly from the command line) independently for each factory.

So one way users could add their own branch is by adding a new factory to `reco.cc`, e.g.

```
app->Add(new JChainMultifactoryGeneratorT<GeneratedJets_factory>(
                  "MyGeneratedJets",
                  {"MCParticles"},
                  {"MyGeneratedJets"},
                  {
                    .jetAlgo="ee_kt_algorithm"
                  },
                  app
));
```

And then recompile and rerun! (I suspect -- but haven't tried it out yet -- you could also do something similar with a plugin processor rather than modifying `reco.cc`...)

**Review by veprbl — APPROVED** [review]

**Review by bspage912 — APPROVED** [review]
Sorry I'm late. I think the functionality is what we need and if Dima is ok with the code and implementation - I'm fine.

**Pull request merged**